File: code/python/data_structures_article-01.py
"""Binary Search Tree Center Method (Features)"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().bstToGst(*args, **kwargs)


class _Solution:
    def bstToGst(self, root: TreeNode) -> TreeNode:
        sum = 0
        def traverse(root: TreeNode) -> None:
            nonlocal sum
            if not root:
                return
            traverse(root.right)
            sum += root.val
            root.val = sum
            traverse(root.left)
        traverse(root)
        return root

File: code/java/data_structures_article-01.java
public class Data_structures_article_01 {
    /**
     * Binary Search Tree Center Method (Features)
     */
    public Object solve(Object... args) {
        if (args == null || args.length != 1) {
            throw new IllegalArgumentException("Expected: (TreeNode root)");
        }

        Solution solution = new Solution();
        return solution.bstToGst((TreeNode) args[0]);
    }

    private static class Solution {

        public TreeNode bstToGst(TreeNode root) {
            traverse(root);
            return root;
        }
        int sum = 0;
        void traverse(TreeNode root) {
            if (root == null) {
                return;
            }
            traverse(root.right);
            sum += root.val;
            root.val = sum;
            traverse(root.left);
        }
    }
}
